utils/rowgroup/rowstorage.h

```diff
@@ -0,0 +1,329 @@
+   * @returns success
+   */
+  bool tryIncreaseInfo();
+
+  /** @brief Reserve space for number of elements (power of two)
+   *
+   *    This function performs re-insert all data
+   *
+   * @param elems(in)   new size
+   */
+  void rehashPowerOfTwo(size_t elems);
+
+  /** @brief Move elements from old one into rehashed data.
+   *
+   *    It's mostly the same algo as in getTargetRow(), but returns nothing
+   *    and skips some checks because it's guaranteed that there is no dups.
+   *
+   * @param oldIdx(in)    index of "old" data
+   * @param oldHashes(in) old storage of row positions and hashes
+   */
+  void insertSwap(size_t oldIdx, RowPosHashStorage * oldHashes);
+
+  /** @brief (Re)Initialize internal data of specified size.
+   *
+   * @param elems(in) number of elements
+   */
+  void initData(size_t elems);
+
+  /** @brief Calculate maximum size of hash assuming 80% fullness.
+   *
+   * @param elems(in) number of elements
+   * @returns calculated size
+   */
+  inline static size_t calcMaxSize(size_t elems) noexcept
+  {
+    if (LIKELY(elems <= std::numeric_limits<size_t>::max() / 100))
+      return elems * 80 / 100;
+
+    return (elems / 100) * 80;
+  }
+
+  inline static size_t calcSizeWithBuffer(size_t elems, size_t maxSize) noexcept
+  {
+    return elems + std::min(maxSize, 0xFFUL);
+  }
+
+  inline static size_t calcSizeWithBuffer(size_t elems) noexcept
+  {
+    return calcSizeWithBuffer(elems, calcMaxSize(elems));
+  }
+
+  /** @brief Calculate memory size of info data
+   *
+   * @param elems(in) number of elements
+   * @returns size in bytes
+   */
+  inline static size_t calcBytes(size_t elems) noexcept
+  {
+    return elems + sizeof(uint64_t);
+  }
+
+  /** @brief Reserve place sufficient for elems
+   *
+   * @param elems(in) number of elements
+   */
+  void reserve(size_t elems);
+
+  /** @brief Start new aggregation generation
+   *
+   * Dump all the data on disk, including internal info data, positions & row
+   * hashes, and the rowgroups itself.
+   */
+  void startNewGeneration();
+
+  /** @brief Save internal info data on disk */
+  void dumpInternalData() const;
+
+  /** @brief Load previously dumped data from disk
+   *
+   * @param gen(in) generation number
+   */
+  void loadGeneration(uint16_t gen);
+  /** @brief Load previously dumped data into the tmp storage */
+  void loadGeneration(uint16_t gen, size_t& size, size_t& mask, size_t& maxSize, uint32_t& infoInc, uint32_t& infoHashShift, uint8_t*& info);
+
+  /** @brief Remove temporary data files */
+  void cleanup();
+  void cleanup(uint16_t gen);
+
+  std::string makeDumpFilename(int32_t gen = -1) const;
+
+private:
+  static constexpr size_t   INIT_SIZE{sizeof(uint64_t)};
+  static constexpr uint32_t INIT_INFO_BITS{5};
+  static constexpr uint8_t  INIT_INFO_INC{1U << INIT_INFO_BITS};
+  static constexpr size_t   INFO_MASK{INIT_INFO_INC - 1U};
+  static constexpr uint8_t  INIT_INFO_HASH_SHIFT{0};
+
+  RowPosHashStorage* fHashes{nullptr};
```

Comment:
Just a note that this initialization breaks default ctags mapping for C++.

Reply:
It seems that universal-ctags is able to work with modern syntax

Reply:
Thx for the note.